src/analyzeFile.ts

```diff
@@ -0,0 +1,95 @@
+import {
+  Conditions,
+  Diagnostic,
+  DiagnosticError,
+  ValidateContext,
+} from "./types";
+import { startsWithKeyword } from "./utils";
+import { after_date } from "./after_date";
+import { on_branch } from "./on_branch";
+import { when } from "./when";
+
+const conditions: Conditions = {
+  after_date,
+  on_branch,
+  when,
+};
+
+function isDiagnosticError(
+  validation: Diagnostic
+): validation is DiagnosticError {
+  return validation.error === true;
+}
+
+export function analyzeFile(
+  file: string,
+  { keywords, options, packageJson }: ValidateContext
+): DiagnosticError[] {
+  const lines = file
+    .split("\n")
+    .map((line, index): [string, number] => [line, index])
```

Comment:
We could improve here a bit the perf by using reduce instead of map and filter.

Reply:
I went for readability over performance. Do you think it will make a big difference in performance? Is it worth it to hand in readability?

Reply:
I would say it is in this case because this is compounded by number of lines in a file and the number of files in a project. for large codebases it will make a difference